File: concepts/08_async_and_concurrency/04_threads_workers_and_process_isolation/lua/test_02_result_and_transfer_boundaries.lua
-- Common question: how do isolated workers transfer results and failures?
-- Inputs: an isolated C state, binary scalar encoding, tables, functions, and userdata.
-- Observations: host-mediated state isolation, explicit value encoding, and non-transferable identities.
-- polyglot-family: async_and_concurrency
-- polyglot-concept: threads_workers_and_process_isolation
-- polyglot-related: languages/lua/tooling_and_runtime/16_advanced_c_api_and_compatibility/
-- polyglot-related+: test_124_structured_data_and_capabilities.lua

local t = require("support.assertions")
local c_api = require("support.c_api")

t.matches(c_api.run("structured"), "C API case passed: structured")

local encoded = string.pack("<i8", 42)
local decoded = string.unpack("<i8", encoded)
t.equal(decoded, 42)

local table_value = {answer = 42}
local function_value = function() return 42 end
local userdata_value = io.tmpfile()
t.equal(type(table_value), "table")
t.equal(type(function_value), "function")
t.equal(type(userdata_value), "userdata")
t.truth(userdata_value:close())

-- Table/function/userdata identity cannot be moved between independent states by a standard Lua API.
t.done()

File: concepts/08_async_and_concurrency/04_threads_workers_and_process_isolation/ruby/test_02_result_and_transfer_boundaries.rb
# 共同问题：隔离 worker 怎样传递结果、所有权和失败。
# 输入：Ractor copy、move、shareable object 和远端异常；观察：身份边界、MovedError 及 RemoteError cause。
# polyglot-family: async_and_concurrency
# polyglot-concept: threads_workers_and_process_isolation
# polyglot-related: languages/ruby/standard_library/11_concurrency_models/
# polyglot-related+: test_088_ractor_move_and_remote_failure.rb

require "assertions"

A = PolyglotAssertions

shareable = Ractor.make_shareable(["ruby"])
A.truth(Ractor.shareable?(shareable))
A.truth(shareable.frozen?)

moved = +"owned"
worker = Ractor.new { Ractor.receive << "-worker" }
worker.send(moved, move: true)
A.equal("owned-worker", worker.value)
A.raises(Ractor::MovedError) { moved.length }

failing = Ractor.new do
  Thread.current.report_on_exception = false
  raise ArgumentError, "remote"
end
remote = A.raises(Ractor::RemoteError) { failing.value }
A.same(ArgumentError, remote.cause.class)
A.equal("remote", remote.cause.message)

A.done
